### PR TITLE
Add encoder/decoder in structureDataset for snowflake.

### PR DESCRIPTION
**Pull request by hhcs9527**
# TL;DR

1. Add encoder/decoder in structureDataset for the snowflake.
2. Currently support the new protocol, snowflake, and sf.
we can add them here, the default is a list of additional protocols, like the following.
image

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Example code for user, we can parse the uro with the following format of uri,
`"snowflake://{dummy_user}:{dummy_account}/{dummy_warehouse}/{dummy_database}/{dummy_schema}/{dummy_table}"`

```python
import mock
import pandas as pd
import pytest
from typing_extensions import Annotated

from flytekit import StructuredDataset, kwtypes, task, workflow, Secret

@task(secret_requests=[Secret(group="snowflake", key="private_key.pem")])
def t1() -> StructuredDataset:
    df = pd.DataFrame({"NAME": ["Tom", "Joseph"], "AGE": [220, 27]})
    return StructuredDataset(
        dataframe=df, uri="sf://phuang9527:rq06467.us-east4.gcp/COMPUTE_WH/AGENT/PEOPLE/PEOPLE"
    )

@task(secret_requests=[Secret(group="snowflake", key="private_key.pem")])
def t2() -> StructuredDataset:
    df = pd.DataFrame({"NAME": ["Tom", "Joseph"], "AGE": [235, 27]})
    return StructuredDataset(
        dataframe=df, uri="snowflake://phuang9527:rq06467.us-east4.gcp/COMPUTE_WH/AGENT/PEOPLE/PEOPLE"
    )

@task(secret_requests=[Secret(group="snowflake", key="private_key.pem")])
def t3(sd: StructuredDataset) -> pd.DataFrame:
    return sd.open(pd.DataFrame).all() # we will load the whole table for user


@workflow
def wf() -> pd.DataFrame:
    t1()
    return t3(sd=t2())

if __name__ == "__main__":
    wf()

```

## Tracking Issue
[Flyte Agent Ecosystem](https://github.com/flyteorg/flyte/issues/3936)

### CODE REVIEW

**Comment by hhcs9527** [comment]
Hi @pingsutw , just finished add ingencoder/decoder in structureDataset for snowflake and unit test, please approve the testing and take a look.

**Comment by hhcs9527** [comment]
Hi @pingsutw, please approve the CI test.

**Comment by hhcs9527** [comment]
Hi @pingsutw, please approve the CI test.

**Comment by hhcs9527** [comment]
Hi @pingsutw, just finished the changes, Please take a look.

**Comment by hhcs9527** [comment]
Hi @pingsutw, just finished the changes to get, please review.
Thanks a lot!

File: flytekit/types/structured/snowflake.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    conn = snowflake.connector.connect(
+        user=user, account=account, private_key=get_private_key(), database=database, schema=schema, warehouse=warehouse
+    )
+
+    cs = conn.cursor()
+    cs.execute(f"select * from {table}")
+
+    dff = cs.fetch_pandas_all()
+    print("cs", cs)
+    print("dff", dff)
```
Comment (pingsutw) [inline flytekit/types/structured/snowflake.py:72]:
```suggestion
```

**Review by pingsutw — APPROVED** [review]

**Comment by hhcs9527** [comment]
Hi @pingsutw, just finished the changes to get, please review.
Thanks a lot!

**Review by pingsutw — DISMISSED** [review]

**Comment by hhcs9527** [comment]
Hi @pingsutw,
I finished the code change to support multiple protocols and fixed the URI.

I will create the new PR to support the new URI.
new URI: `uri="snowflake://dummy_user:dummy_account/dummy_warehouse/dummy_database/dummy_schema/dummy_table`

Thanks a lot

**Comment by Future-Outlier** [comment]
This is done by other PR, so I'm going to close it, thank you @hhcs9527 !
